File: app/src/main/java/com/example/ifoundit/MainActivity.java
package com.example.ifoundit;

import androidx.annotation.NonNull;
import androidx.annotation.UiThread;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import com.naver.maps.map.MapView;
import com.naver.maps.map.NaverMap;
import com.naver.maps.map.NaverMapSdk;
import com.naver.maps.map.OnMapReadyCallback;

public class MainActivity extends AppCompatActivity implements OnMapReadyCallback {
    MapView mapView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        init();

        //NaverMap 객체 얻어오기
        getMapInstance();

    }

    void init(){
        mapView = findViewById(R.id.mapView);
    }

    void getMapInstance(){
        //인증 실패 처리
        /*
         * UnauthorizedClientException(401):잘못된 클라이언트 ID를 지정함, 잘못된 클라이언트 유형을 사용함, 콘솔에서 앱 패키지 이름을 잘못 등록함
         * QuotaExceededException(429):콘솔에서 Maps 서비스를 선택하지 않음, 사용 한도가 초과됨
         * ClientUnspecifiedException(800):클라이언트 ID를 지정하지 않음
         * */
        NaverMapSdk.getInstance(this).setOnAuthFailedListener(new NaverMapSdk.OnAuthFailedListener() {
            @Override
            public void onAuthFailed(@NonNull NaverMapSdk.AuthFailedException e) {
                e.printStackTrace();
            }
        });

        //NaverMap 객체 얻어오기 (API 호출 가능) -> onMapReady 콜백 메서드 호출
        mapView.getMapAsync(this);
    }

    @UiThread
    @Override
    public void onMapReady(@NonNull NaverMap naverMap) {

    }
}

File: EventMap/app/src/main/java/com/example/android/eventmap/view/main/MainActivity.kt
package com.example.android.eventmap.view.main

import android.os.Bundle
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.lifecycleScope
import com.example.android.eventmap.R
import com.example.android.eventmap.databinding.ActivityMainBinding
import com.example.android.eventmap.util.extensions.toastMessage
import com.example.android.eventmap.view.main.type.MapKindType
import com.naver.maps.map.MapFragment
import com.naver.maps.map.NaverMap
import com.naver.maps.map.OnMapReadyCallback
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

/**
 * Created by Jihye Noh
 * Date: 2022-01-09
 */
@AndroidEntryPoint
class MainActivity : AppCompatActivity(), OnMapReadyCallback {
    private val binding: ActivityMainBinding by lazy {
        DataBindingUtil.setContentView(this, R.layout.activity_main)
    }

    private val viewModel by viewModels<MainViewModel>()

    private var naverMap: NaverMap? = null
    private var mapView = MapFragment()
    private var backPressedTime: Long = 0L

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding.apply {
            model = viewModel
            lifecycleOwner = this@MainActivity
        }

        initNaverMapInstance()
        collectFlows()
    }

    override fun onDestroy() {
        super.onDestroy()
        naverMap = null
    }

    override fun onBackPressed() {
        closeNaviMapKindDrawable()
    }

    private fun initNaverMapInstance() {
        mapView.getMapAsync(this)
    }

    override fun onMapReady(map: NaverMap) {
        naverMap = map
    }

    private fun collectFlows() {
        lifecycleScope.launch {
            viewModel.eventFlow.collect { event ->
                handleEvent(event)
            }
        }
    }

    private fun handleEvent(event: MainViewModel.Event) = when (event) {
        is MainViewModel.Event.MapSettingOpenEvent -> {
            binding.mainDrawerLayout.openDrawer(GravityCompat.START)
        }
        is MainViewModel.Event.KindOfMapClickEvent -> {
            when (event.kind) {
                MapKindType.BASIC -> {
                    naverMap?.mapType = NaverMap.MapType.Basic
                }
                MapKindType.SATELLITE -> {
                    naverMap?.mapType = NaverMap.MapType.Satellite
                }
                MapKindType.TERRAIN -> {
                    naverMap?.mapType = NaverMap.MapType.Terrain
                }
                MapKindType.NAVI -> {
                    naverMap?.mapType = NaverMap.MapType.Navi
                }
            }
        }
    }

    private fun closeNaviMapKindDrawable() {
        if (binding.mainDrawerLayout.isDrawerOpen(binding.layoutNavigationheaderSetting.navigationviewSetting)) {
            binding.mainDrawerLayout.closeDrawer(GravityCompat.START)
            return
        }

        val currentTime = System.currentTimeMillis()
        when (currentTime - backPressedTime) {
            in 0..2000 -> {
                super.onBackPressed()
            }
            else -> {
                backPressedTime = currentTime
                toastMessage(R.string.one_press_back_button_toast)
            }
        }
    }
}